Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,260 @@
+        :param tuple:
+        :return: any type
+        """
+        row = tuple[0]
+        col = tuple[1]
+        return self.grid[row][col]
+
+    def findElementInGrid(self, element):
+        """
+        Return the list of all Points with the grid and coordinates of the element if exists. None otherwise.
+        :param element: any type
+        :return: Point
+        """
+        result = []
+        numRows, numCols = self.getSize()
+        for i in range(numRows):
+            for j in range(numCols):
+                if self.grid[i][j] == element:
+                    result.append(Point(self, (i, j), element))
+        return result
+
+    def findNeighborsOfPoint(self, tuple):
+        """
+        Return list of Points that are neighbours of element with the given coordinates.
+        :param tuple: Tuple
+        :return: list(Point)
+        """
+        row = tuple[0]
+        col = tuple[1]
+        result = []
+        possible_coord = [(row-1, col-1),(row-1, col),(row, col-1),(row-1, col+1),
+                          (row+1, col-1),(row+1, col),(row, col+1),(row+1, col+1)]
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result
+    
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        if self.findString(word) is None:
+            return False
+        return True
+    
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        bub can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: bool
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints: # we go further only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        return [p.value for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):
+        """
+        Helper-function to find a word in a grid. Finds the longest path to form the word (may be only a part)
+        :param point: Point
+        :param word: str
+        :param path: list(Point)
+        :param visited: list(Point)
+        :return: None
+        """
+        if self._pathIsWord(path, word):
+            return path
+
+        possible_ways = self.findPossibleWays(path, point, word, visited)
+        if possible_ways == []:

```

Comment:
It looks to me like this should be a terminal condition, the method should return `None` as nothing was found in this tree.